internal/tests/cluster/session_cleanup_test.go

```diff
@@ -0,0 +1,513 @@
+	"github.com/hashicorp/vault/sdk/helper/base62"
+	"github.com/mr-tron/base58"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/protobuf/proto"
+	"nhooyr.io/websocket"
+	"nhooyr.io/websocket/wspb"
+)
+
+const testSendRecvSendMax = 60
+
+func TestWorkerSessionCleanup(t *testing.T) {
+	require := require.New(t)
+	logger := hclog.New(&hclog.LoggerOptions{
+		Level: hclog.Trace,
+	})
+
+	conf, err := config.DevController()
+	require.NoError(err)
+
+	pl, err := net.Listen("tcp", "localhost:0")
+	require.NoError(err)
+	c1 := controller.NewTestController(t, &controller.TestControllerOpts{
+		Config:                 conf,
+		InitialResourcesSuffix: "1234567890",
+		Logger:                 logger.Named("c1"),
+		PublicClusterAddr:      pl.Addr().String(),
+	})
+	defer c1.Shutdown()
+
+	expectWorkers(t, c1)
+
+	// Wire up the testing proxies
+	require.Len(c1.ClusterAddrs(), 1)
+	proxy, err := dawdle.NewProxy("tcp", "", c1.ClusterAddrs()[0],
+		dawdle.WithListener(pl),
+		dawdle.WithRbufSize(512),
+		dawdle.WithWbufSize(512),
+	)
+	require.NoError(err)
+	defer proxy.Close()
+	require.NotEmpty(t, proxy.ListenerAddr())
+
+	w1 := worker.NewTestWorker(t, &worker.TestWorkerOpts{
+		WorkerAuthKms:      c1.Config().WorkerAuthKms,
+		InitialControllers: []string{proxy.ListenerAddr()},
+		Logger:             logger.Named("w1"),
+	})
+	defer w1.Shutdown()
+
+	time.Sleep(10 * time.Second)
+	expectWorkers(t, c1, w1)
+
+	// Use an independent context for test things that take a context so
+	// that we aren't tied to any timeouts in the controller, etc. This
+	// can interfere with some of the test operations.
+	ctx := context.Background()
+
+	// Connect target
+	client := c1.Client()
+	client.SetToken(c1.Token().Token)
+	tcl := targets.NewClient(client)
+	tgt, err := tcl.Read(ctx, "ttcp_1234567890")
+	require.NoError(err)
+	require.NotNil(tgt)
+
+	// Create test server, update default port on target
+	ts := newTestTcpServer(t, logger)
+	require.NotNil(t, ts)
+	defer ts.Close()
+	tgt, err = tcl.Update(ctx, tgt.Item.Id, tgt.Item.Version, targets.WithTcpTargetDefaultPort(ts.Port()))
+	require.NoError(err)
+	require.NotNil(tgt)
+
+	// Authorize and connect
+	sess := newTestSession(ctx, t, tcl, "ttcp_1234567890")
+	sConn := sess.Connect(ctx, t, logger)
+
+	// Run initial send/receive test, make sure things are working
+	sConn.TestSendRecvAll(t)
+
+	// Kill the link
+	proxy.Pause()
+
+	// Run again, ensure connection is dead
+	sConn.TestSendRecvFail(t)
+
+	// Assert we have no connections left (should be default behavior)
+	sess.TestNoConnectionsLeft(t)
+
+	// Resume the connection, and reconnect.
+	proxy.Resume()
+	time.Sleep(time.Second * 10)                          // Sleep to wait for worker to report back as healthy
+	sess = newTestSession(ctx, t, tcl, "ttcp_1234567890") // re-assign, other connection will close in t.Cleanup()
+	sConn = sess.Connect(ctx, t, logger)
+	sConn.TestSendRecvAll(t)
+}
+
+func TestWorkerSessionCleanupMultiController(t *testing.T) {
```

Comment:
I'm also confused about what non-test fixture is under test.

Reply:
Discussed this offline with @jimlambrt and the confusion was due to the use of the `TestConnection` and `TestWorker` helpers. This should be cleared up now with the discussion.